modules/dev-utils/src/main/java/org/apache/ignite/development/utils/IgniteIndexReader.java

```diff
@@ -0,0 +1,2130 @@
+            ", cacheId=" + cacheAwareLink.cacheId + ", partId=" + partId +
+            ", pageIndex=" + pageIdx + ", itemId=" + itemId + ", link=" + link;
+    }
+
+    /**
+     * Finds certain pages in file page store. When all pages corresponding given types is found, at least one page
+     * for each type, we return the result.
+     *
+     * @param partId Partition id.
+     * @param flag Page store flag.
+     * @param store File page store.
+     * @param pageTypes Page types to find.
+     * @return Map of found pages. First page of this class that was found, is put to this map.
+     * @throws IgniteCheckedException If failed.
+     */
+    private Map<Class, Long> findPages(int partId, byte flag, FilePageStore store, Set<Class> pageTypes)
+        throws IgniteCheckedException {
+        Map<Class, Long> res = new HashMap<>();
+
+        scanFileStore(partId, flag, store, (pageId, addr, io) -> {
+            if (pageTypes.contains(io.getClass())) {
+                res.put(io.getClass(), pageId);
+
+                pageTypes.remove(io.getClass());
+            }
+
+            return !pageTypes.isEmpty();
+        });
+
+        return res;
+    }
+
+    /**
+     * Compares result of traversals.
+     *
+     * @param treeInfos Traversal from root to leafs.
+     * @param treeScans Traversal using horizontal scan.
+     */
+    private void compareTraversals(Map<String, TreeTraversalInfo> treeInfos, Map<String, TreeTraversalInfo> treeScans) {
+        List<String> errors = new LinkedList<>();
+
+        Set<String> treeIdxNames = new HashSet<>();
+
+        treeInfos.forEach((name, tree) -> {
+            treeIdxNames.add(name);
+
+            TreeTraversalInfo scan = treeScans.get(name);
+
+            if (scan == null) {
+                errors.add("Tree was detected in " + RECURSIVE_TRAVERSE_NAME + " but absent in  "
+                    + HORIZONTAL_SCAN_NAME + ": " + name);
+
+                return;
+            }
+
+            if (tree.itemStorage.size() != scan.itemStorage.size())
+                errors.add(compareError("items", name, tree.itemStorage.size(), scan.itemStorage.size(), null));
+
+            Set<Class> classesInStat = new HashSet<>();
+
+            tree.ioStat.forEach((cls, cnt) -> {
+                classesInStat.add(cls);
+
+                AtomicLong scanCnt = scan.ioStat.get(cls);
+
+                if (scanCnt == null)
+                    scanCnt = new AtomicLong(0);
+
+                if (scanCnt.get() != cnt.get())
+                    errors.add(compareError("pages", name, cnt.get(), scanCnt.get(), cls));
+            });
+
+            scan.ioStat.forEach((cls, cnt) -> {
+                if (classesInStat.contains(cls))
+                    // Already checked.
+                    return;
+
+                errors.add(compareError("pages", name, 0, cnt.get(), cls));
+            });
+        });
+
+        treeScans.forEach((name, tree) -> {
+            if (!treeIdxNames.contains(name))
```

Comment:
Are you sure checking _key_PK correct?

Reply:
Yes, because tree names for _key_PK indexes are different as they include cache id and type id.